src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,204 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    internal sealed class TranscodingWriteStream : Stream
+    {
+        internal const int MaxCharBufferSize = 4096;
+        internal const int MaxByteBufferSize = 4 * MaxCharBufferSize;
+        private readonly int _maxByteBufferSize;
+
+        private readonly Stream _stream;
+        private readonly Decoder _decoder;
+        private readonly Encoder _encoder;
+        private readonly char[] _charBuffer;
+        private int _charsDecoded;
+        private bool _disposed;
+
+        public TranscodingWriteStream(Stream stream, Encoding targetEncoding)
+        {
+            _stream = stream;
+
+            _charBuffer = ArrayPool<char>.Shared.Rent(MaxCharBufferSize);
+
+            // Attempt to allocate a byte buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the char -> byte conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            _maxByteBufferSize = Math.Min(MaxByteBufferSize, targetEncoding.GetMaxByteCount(MaxCharBufferSize));
+
+            _decoder = Encoding.UTF8.GetDecoder();
+            _encoder = targetEncoding.GetEncoder();
+        }
+
+        public override bool CanRead => false;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get; set; }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override async Task FlushAsync(CancellationToken cancellationToken)
+        {
+            await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
```

Comment:
Write should be implemented, as should Begin/EndWrite, WriteByte, Write(Span), WriteAsync(ReadOnlyMemory), etc.

Reply:
@ManickaP is currently implementing sync APIs for HttpClient (https://github.com/dotnet/runtime/issues/32125).  Once that's in place, these synchronous Write/Flush methods will become callable.  If we're not exposing this stream publicly, then technically we only need to implement the sync APIs that HttpClient itself will use, which means Write/Flush/etc.  You and she can work out whether it makes to do that in your PR or hers, but we should make sure it happens, otherwise this Content type will be broken.

Reply:
Same commnet as here https://github.com/dotnet/runtime/pull/33459#discussion_r395833435